File: algorithms/medium/sort_vowels_in_a_string.ex
defmodule Solution do
  @spec sort_vowels(s :: String.t()) :: String.t()
  def sort_vowels(s) do
    cs = String.graphemes(s)

    vowels = Enum.filter(cs, &is_vowel?/1) |> Enum.sort()

    cs
    |> Enum.reduce({[], vowels}, fn c, {ans, vs} ->
      if is_vowel?(c) do
        [v | vs] = vs
        {[v | ans], vs}
      else
        {[c | ans], vs}
      end
    end)
    |> elem(0)
    |> Enum.reverse()
    |> Enum.join("")
  end

  defp is_vowel?(c) do
    String.downcase(c) in ["a", "e", "i", "o", "u"]
  end
end

File: algorithms/medium/sort_vowels_in_a_string.rs
impl Solution {
    pub fn sort_vowels(s: String) -> String {
        let mut vowels: Vec<u8> = s.bytes().filter(is_vowel).collect();

        vowels.sort_unstable();

        let mut s = s.into_bytes();

        let mut vowels = vowels.into_iter();

        for c in &mut s {
            if is_vowel(c) {
                *c = vowels.next().unwrap();
            }
        }

        unsafe { String::from_utf8_unchecked(s) }
    }
}

fn is_vowel(&c: &u8) -> bool {
    match c {
        b'a' | b'e' | b'i' | b'o' | b'u' => true,
        b'A' | b'E' | b'I' | b'O' | b'U' => true,
        _ => false,
    }
}
